### ISSUE TITLE
code

### ISSUE THREAD

**Issue by kvshilpa610** [kvshilpa610/pdf-chatbot#8]
import os
import re
import numpy as np
from pypdf import PdfReader

import faiss
from rank_bm25 import BM25Okapi

from sentence_transformers import SentenceTransformer, CrossEncoder

from transformers import AutoTokenizer, AutoModelForSeq2SeqLM
import torch

# ============================================================
# 1) PDF TEXT EXTRACTION
# ============================================================
def extract_text_from_pdf(pdf_path: str) -> str:
    reader = PdfReader(pdf_path)
    all_text = []

    for i, page in enumerate(reader.pages):
        page_text = page.extract_text() or ""
        page_text = page_text.strip()

        if page_text:
            all_text.append(f"\n\n--- PAGE {i+1} ---\n{page_text}")

    return "\n".join(all_text)

# ============================================================
# 2) CLEAN TEXT
# ============================================================
def clean_text(text: str) -> str:
    text = text.replace("\t", " ")
    text = re.sub(r"[ ]{2,}", " ", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip()

# ============================================================
# 3) CHUNKING
# ============================================================
def chunk_text(text: str, chunk_size=1000, overlap=200):
    chunks = []
    start = 0

    while start < len(text):
        end = start + chunk_size
        chunk = text[start:end].strip()

        if len(chunk) > 200:
            chunks.append(chunk)

        start = end - overlap

    return chunks

# ============================================================
# 4) TOKENIZE FOR BM25
# ============================================================
def bm25_tokenize(text: str):
    text = text.lower()
    text = re.sub(r"[^a-z0-9\s]", " ", text)
    return [w for w in text.split() if len(w) > 2]

# ============================================================
# 5) BUILD FAISS INDEX
# ============================================================
def build_faiss_index(embeddings: np.ndarray):
    dim = embeddings.shape[1]
    index = faiss.IndexFlatIP(dim)
    index.add(embeddings)
    return index

# ============================================================
# 6) HYBRID RETRIEVAL (SAFE VERSION)
# ============================================================
def hybrid_retrieve(query, chunks, faiss_index, embed_model, bm25, top_k=10):
    
    # ----- VECTOR SEARCH -----
    q_emb = embed_model.encode([query], normalize_embeddings=True).astype("float32")
    scores_vec, ids_vec = faiss_index.search(q_emb, top_k)

    vec_results = [
        (int(idx), float(score))
        for score, idx in zip(scores_vec[0], ids_vec[0])
        if idx != -1
    ]

    # ----- BM25 SEARCH -----
    tokenized_query = bm25_tokenize(query)
    bm25_scores = bm25.get_scores(tokenized_query)

    bm25_top_ids = np.argsort(bm25_scores)[::-1][:top_k]
    bm25_results = [(int(i), float(bm25_scores[i])) for i in bm25_top_ids]

    bm25_vals = [s for _, s in bm25_results]
    max_bm25 = max(bm25_vals) if bm25_vals else 0.0

    combined = {}

    # Add vector scores
    for idx, s in vec_results:
        combined[idx] = combined.get(idx, 0) + (s * 1.0)

    # Add BM25 scores safely
    for idx, s in bm25_results:
        if max_bm25 > 0:
            normalized_score = s / max_bm25
        else:
            normalized_score = 0.0

        combined[idx] = combined.get(idx, 0) + (normalized_score * 0.9)

    ranked = sorted(combined.items(), key=lambda x: x[1], reverse=True)
    final_ids = [idx for idx, _ in ranked[:top_k]]

    return final_ids

# ============================================================
# 7) RERANKING
# ============================================================
def rerank(query, retrieved_chunks, reranker, top_n=4):
    pairs = [(query, c) for c in retrieved_chunks]
    scores = reranker.predict(pairs)

    ranked = sorted(
        list(zip(retrieved_chunks, scores)),
        key=lambda x: x[1],
        reverse=True
    )

    return [c for c, s in ranked[:top_n]]

# ============================================================
# 8) ANSWER USING FLAN-T5
# ============================================================
def answer_with_flan_t5(query, best_chunks, tokenizer, model, device):
    context = "\n\n".join(best_chunks)

    prompt = f"""
Answer the question ONLY using the context.
If not found, say: Answer not found in the PDF.

Context:
{context}

Question:
{query}

Answer in 8 to 12 bullet points in simple words.
End with 2-line summary.
"""

    inputs = tokenizer(
        prompt,
        return_tensors="pt",
        truncation=True,
        max_length=1024
    ).to(device)

    with torch.no_grad():
        outputs = model.generate(
            **inputs,
            max_new_tokens=220,
            num_beams=4,
            do_sample=False
        )

    return tokenizer.decode(outputs[0], skip_special_tokens=True)

# ============================================================
# MAIN PROGRAM
# ============================================================
def main():
    print("\n=================================================")
    print("📄 PDF RAG CHATBOT (100% FREE, NO OLLAMA)")
    print("Hybrid Retrieval + Reranking + FLAN-T5 Answering")
    print("=================================================\n")

    pdf_path = r"C:\Users\shilpa\OneDrive\Desktop\sample.pdf"

    if not os.path.exists(pdf_path):
        print("\n❌ PDF file not found.\n")
        return

    print("\n✅ Reading PDF...")
    raw_text = extract_text_from_pdf(pdf_path)

    if not raw_text.strip():
        print("\n❌ No text extracted. PDF may be scanned.\n")
        return

    print("✅ Cleaning text...")
    raw_text = clean_text(raw_text)

    print("✅ Chunking...")
    chunks = chunk_text(raw_text, chunk_size=1000, overlap=200)
    print(f"✅ Total chunks: {len(chunks)}")

    print("\n✅ Loading embedding model...")
    embed_model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2")

    print("✅ Creating embeddings...")
    embeddings = embed_model.encode(
        chunks,
        normalize_embeddings=True
    ).astype("float32")

    print("✅ Building FAISS...")
    faiss_index = build_faiss_index(embeddings)

    print("✅ Building BM25...")
    tokenized_chunks = [bm25_tokenize(c) for c in chunks]
    bm25 = BM25Okapi(tokenized_chunks)

    print("\n✅ Loading reranker...")
    reranker = CrossEncoder("cross-encoder/ms-marco-MiniLM-L-6-v2")

    print("\n✅ Loading FREE LLM model (FLAN-T5)...")
    model_name = "google/flan-t5-base"

    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForSeq2SeqLM.from_pretrained(model_name)

    device = "cuda" if torch.cuda.is_available() else "cpu"
    model = model.to(device)

    print(f"✅ Running on: {device}")

    print("\n🎉 Chatbot ready!")
    print("Type: exit (to stop)\n")

    while True:
        query = input("You: ").strip()

        if query.lower() in ["exit", "quit", "stop"]:
            print("\n👋 Bye Shilpa!\n")
            break

        ids = hybrid_retrieve(query, chunks, faiss_index, embed_model, bm25, top_k=10)
        retrieved = [chunks[i] for i in ids]

        best_chunks = rerank(query, retrieved, reranker, top_n=4)

        answer = answer_with_flan_t5(query, best_chunks, tokenizer, model, device)

        print("\n🤖 Answer:\n")
        print(answer)
        print("\n" + "-" * 60 + "\n")

if __name__ == "__main__":
    main()
